### PR TITLE
Add rounding by decimal placement

### PR DESCRIPTION
**Pull request by TheLimeGlass**
### Description
Add rounding by decimal placement to the round function.

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** https://github.com/SkriptLang/Skript/issues/3579

### CODE REVIEW

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 1]
```diff
@@ -68,16 +64,21 @@ public Long[] executeSimple(Object[][] params) {
...
-		Functions.registerFunction(new SimpleJavaFunction<Long>("round", numberParam, DefaultClasses.LONG, true) {
+		Functions.registerFunction(new SimpleJavaFunction<Number>("round", new Parameter[] {new Parameter<>("n", DefaultClasses.NUMBER, true, null), new Parameter<>("d", DefaultClasses.NUMBER, true, new SimpleLiteral<Number>(-1, false))}, DefaultClasses.NUMBER, true) {
 			@Override
-			public Long[] executeSimple(Object[][] params) {
+			public Number[] executeSimple(Object[][] params) {
 				if (params[0][0] instanceof Long)
 					return new Long[] {(Long) params[0][0]};
-				return new Long[] {Math2.round(((Number) params[0][0]).doubleValue())};
+				double value = ((Number) params[0][0]).doubleValue();
+				int placement = ((Number)params[1][0]).intValue();
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:73]:
```suggestion
				int placement = ((Number) params[1][0]).intValue();
```

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 2]
```diff
@@ -68,16 +64,21 @@ public Long[] executeSimple(Object[][] params) {
...
+				if (placement > 0) {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:74]: You can omit braces here (optional)

**Review by AyhamAl-Ali — APPROVED** [review]
Noice addition ⚡

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java (hunk 2 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:74]: Can't as the `params` array is an Object

**Review by AyhamAl-Ali — APPROVED** [review]

**Comment by TPGamesNL** [comment]
I think it would be nice if negative numbers were allowed to, e.g. `123` rounded to -1 decimal points is `120`

**Comment by TheLimeGlass** [comment]
> I think it would be nice if negative numbers were allowed to, e.g. `123` rounded to -1 decimal points is `120`

I don't think that'll be a good addition, because the number acts as an index.

Based on your suggest, I understood that you want to grab the current decimal placements subtracted by that number. Is that your suggestion?

**Comment by TPGamesNL** [comment]
My suggestion is to extend the index to also allow it to index into the part to the left of the decimal separator. 
For example:
```
1234.5678
    ^      index  0: 1235.0000
      ^    index  2: 1234.5700
       ^   index  3: 1234.5680
  ^        index -2: 1200.000
```

**Comment by TheLimeGlass** [comment]
> My suggestion is to extend the index to also allow it to index into the part to the left of the decimal separator. For example:
> 
> ```
> 1234.5678
>     ^      index  0: 1235.0000
>       ^    index  2: 1234.5700
>        ^   index  3: 1234.5680
>   ^        index -2: 1200.000
> ```

That's now how mathematics work though.

So `round(15.1, -1)` is 2.1?

**Comment by TPGamesNL** [comment]
No, `round(15.5, -1)` is `20.0`

**Comment by jayjna** [comment]
I like this idea

**Comment by Fusezion** [comment]
based off test does this mean it bypasses the number decimal limit in config?

**Comment by TheLimeGlass** [comment]
> based off test does this mean it bypasses the number decimal limit in config?

It says that's for displaying.

I don't think that setting should be considered here. If they're defining the amount of decimal places, they would already know the placement they want, and it should override that setting.

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 3]
```diff
@@ -68,16 +66,27 @@ public Long[] executeSimple(Object[][] params) {
...
-			public Long[] executeSimple(Object[][] params) {
+			public Number[] executeSimple(Object[][] params) {
 				if (params[0][0] instanceof Long)
 					return new Long[] {(Long) params[0][0]};
-				return new Long[] {Math2.round(((Number) params[0][0]).doubleValue())};
+				double value = ((Number) params[0][0]).doubleValue();
+				int placement = ((Number) params[1][0]).intValue();
+				if (placement != 0) {
+					if (placement >= 0) {
+						BigDecimal decimal = new BigDecimal(Double.toString(value));
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:78]: why are you using BigDecimal here? can't you just do it with a normal double and a bit of math

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 4]
```diff
@@ -68,16 +66,27 @@ public Long[] executeSimple(Object[][] params) {
...
+		Functions.registerFunction(new SimpleJavaFunction<Number>("round", new Parameter[] {new Parameter<>("n", DefaultClasses.NUMBER, true, null), new Parameter<>("d", DefaultClasses.NUMBER, true, new SimpleLiteral<Number>(0, false))}, DefaultClasses.NUMBER, true) {
 			@Override
-			public Long[] executeSimple(Object[][] params) {
+			public Number[] executeSimple(Object[][] params) {
 				if (params[0][0] instanceof Long)
 					return new Long[] {(Long) params[0][0]};
-				return new Long[] {Math2.round(((Number) params[0][0]).doubleValue())};
+				double value = ((Number) params[0][0]).doubleValue();
+				int placement = ((Number) params[1][0]).intValue();
+				if (placement != 0) {
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:76]: the if statements could be structured like this to reduce nesting
```
if (placement == 0) {

} else if (placement > 0) {

} else {

}
```

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java (hunk 3 above)
Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:78]: It's for the utility, the methods the class provides are used below.

**Review by UnderscoreTud — APPROVED** [review]

**Pull request merged**